File: src/main/java/com/zippeykeys/praisebe/pattern/ILocalize.java
package com.zippeykeys.praisebe.pattern;

import com.zippeykeys.praisebe.Reference;
import com.zippeykeys.praisebe.util.Util;

import org.jetbrains.annotations.Contract;

import net.minecraft.util.IStringSerializable;
import net.minecraft.util.ResourceLocation;

public interface ILocalize extends IStringSerializable {
    @Override
    String getName();

    String getPrefix();

    @Contract(pure = true)
    default String getUnlocalizedName() {
        return getPrefix() + "." + getName();
    }

    @Contract(pure = true)
    default String getLocalizedName() {
        return Reference.INSTANCE.proxy().localize(getUnlocalizedName() + ".name");
    }

    @Contract(pure = true)
    default String getLocalizedDescription() {
        return Reference.INSTANCE.proxy().localize(getUnlocalizedName() + ".desc");
    }

    default ResourceLocation getResource() {
        return Util.getResource(getName());
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/iface/ILocalize.kt
package com.zippeykeys.praisebe.iface

import com.zippeykeys.praisebe.Reference
import com.zippeykeys.praisebe.util.getResource
import net.minecraft.util.IStringSerializable
import net.minecraft.util.ResourceLocation


interface ILocalize : IStringSerializable {
    val prefix: String

    val resource: ResourceLocation
        get() = getResource(name)

    fun getUnlocalizedName(): String {
        return "$prefix.$name"
    }

    fun getLocalizedName(): String {
        return Reference.INSTANCE.proxy()
                .localize(getUnlocalizedName() + ".name")
    }

    fun getLocalizedDescription(): String {
        return Reference.INSTANCE.proxy()
                .localize(getUnlocalizedName() + ".desc")
    }
}